### DISCUSSION TITLE
Descriptors, *args, and **kwargs?

### DISCUSSION THREAD

**Discussion by sad2project** [exaloop/codon#568] [General]
I'm curious if this supports Python's descriptors and the use of *args and **kwargs.
Based on your restrictions for tuples, I'm guessing the latter is out, but I can't tell if the lack of descriptors being mentioned is a sign of not being included in the feature set or if you just haven't mentioned it in the docs yet. Even if you don't fully support descriptors, I think support of @property would be good.

**Comment by arshajii** [comment]
Hi @sad2project -- yes Codon supports all of these (I'm assuming you mean *decorator* as opposed to *descriptor*?):

``` python
# compiles in Codon

def mydecorator(func):
    def wrapper(*args, **kwargs):
        print('calling function!')
        result = func(*args, **kwargs)
        return result
    return wrapper

@mydecorator
def foo(a, b, c, **kwargs):
    return a + b + c + len(kwargs)

print(foo(12, 3.14, 99, x='hello', y=[1,2,3]))
```

There are a few limitations since we do this all statically at compile time, but for the most part it works just like in Python.

**Comment by sad2project** [comment]
I can see why you thought I meant decorator, but I meant descriptor
